src/IO/H5/VolumeData.cpp

```diff
@@ -159,6 +165,335 @@ void append_element_extents_and_connectivity(
+// std::vector is the list of all such grid points
+template <size_t SpatialDim>
+std::vector<std::array<double, SpatialDim>> generate_block_logical_coordinates(
+    const std::vector<std::array<double, SpatialDim>>&
+        element_logical_coordinates,
+    const std::string& grid_name,
+    const std::array<int, SpatialDim>& h_refinement_array) {
+  size_t grid_points_x_start_position = 0;
+  std::vector<std::array<double, SpatialDim>> block_logical_coordinates;
+  block_logical_coordinates.reserve(element_logical_coordinates.size());
+  std::vector<double> number_of_elements_each_direction;
+  number_of_elements_each_direction.reserve(SpatialDim);
+  std::vector<double> shift_each_direction;
+  shift_each_direction.reserve(SpatialDim);
+
+  // Computes number_of_elements_each_direction, element_index, and
+  // shift_each_direction to each be used in the computation of the grid point
+  // coordinates in the block logical frame
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    double number_of_elements = pow(2, gsl::at(h_refinement_array, i));
+    number_of_elements_each_direction.push_back(number_of_elements);
+    size_t grid_points_start_position =
+        grid_name.find('I', grid_points_x_start_position + 1);
+    size_t grid_points_end_position =
+        grid_name.find(',', grid_points_start_position);
+    if (i == SpatialDim) {
+      grid_points_end_position =
+          grid_name.find(')', grid_points_start_position);
+    }
+    int element_index = std::stoi(grid_name.substr(
+        grid_points_start_position + 1,
+        grid_points_end_position - grid_points_start_position - 1));
+    double shift = (-1 + (2 * element_index + 1) / number_of_elements);
+    shift_each_direction.push_back(shift);
+    grid_points_x_start_position = grid_points_start_position;
+  }
+
+  // Computes the coordinates for each grid point in the block logical frame
+  for (size_t i = 0; i < element_logical_coordinates.size(); ++i) {
+    std::array<double, SpatialDim> grid_point_coordinate = {};
+    for (size_t j = 0; j < grid_point_coordinate.size(); ++j) {
+      gsl::at(grid_point_coordinate, j) =
+          1. / number_of_elements_each_direction[j] *
+              element_logical_coordinates[i][j] +
+          shift_each_direction[j];
+    }
+    block_logical_coordinates.push_back(grid_point_coordinate);
+  }
+
+  return block_logical_coordinates;
+}
+
+// Given a std::vector<double> where the elements are ordered in ascending order
+// a new std::vector<double> is generated where it is a list of the original
+// values in ascending order without duplicates
+// Example: [1,2,2,3] -> order() -> [1,2,3]
+std::vector<double> order_sorted_elements(
+    const std::vector<double>& sorted_elements) {
+  std::vector<double> ordered_elements;
+  ordered_elements.push_back(sorted_elements[0]);
+  for (size_t i = 1; i < sorted_elements.size(); ++i) {
+    if (sorted_elements[i] == ordered_elements.end()[-1]) {
```

Comment:
What about `if (... != ...)` and then put the `else` part in there. The continue doesn't really do anything. Normally continue is used to skip a remaining bit of work in a loop, but there's no addition work inside this loop